# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [ ]:
(lambda x:x+2)(2)
a_var=4
print((lambda x:x+2)(a_var))

6


In [ ]:
add_two=lambda x:x+2
add_two(2)

4

#### Saving to a Variable

In [ ]:
add_two=lambda x:x+2
add_two(2)



f_test=lambda x:x+2
f_test(4)

6

#### Multiple Inputs

In [ ]:
add_two=lambda x,y,z:x**2+y**2+z**2
add_two(2,3,4)

29

#### Passing a Lambda into a Function

In [ ]:
def multiply_two(f,num):
  return (lambda x,y:x*y)(f,num)


multiply_two(12,3)



36

#### Returning a Lambda from a Function

In [ ]:
def a_func(b,c):# what value r for what
  return (lambda x,a:a+b+c)(5,5)

a_func(5,5)


15

#### If Statements within Lambdas

In [ ]:
(lambda x,y:True if x>y else False) (20,45)
#(lambda x,y:x+2 if x>y else x-y) (100,50)

False

In [ ]:
f_condition=lambda x:x+2 if x<10 else x+10
print(f_condition(10))
f_condition(2)
f_condition(30)

40

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [ ]:
f=(lambda x:x**3 )(2)
f

8

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [ ]:
c=lambda b:(b-1,b+3,b*2)
a=lambda x:x+2
list(map(c,[1,2,3,4,5]))


list(map(a,[1,2,3,3]))

[3, 4, 5, 5]

In [ ]:
def squared(num,num_2):
  if num<100 and num_2<200:
    return num**2 , num_2**2
  else :
    return num, num_2

numbers=[4,5,3,6,7,8,1,89,67,900,60,89]
more_numbers=[4,2,5,6,3,89,56]
squared_nums_map=list(map(squared,numbers,more_numbers))
print(squared_nums_map)


[(16, 16), (25, 4), (9, 25), (36, 36), (49, 9), (64, 7921), (1, 3136)]


#### Using Lambda's with Map

In [ ]:
add_three=lambda x:(x*2)
list(map(add_three,[1,2,3,4,5]))

[2, 4, 6, 8, 10]

In [ ]:
squared_num=list(map(lambda x,y:(x**2,y**2) if x<y else (x,y),numbers,more_numbers))
print(squared_num)

[(4, 4), (5, 2), (9, 25), (6, 6), (7, 3), (64, 7921), (1, 3136)]


In [ ]:
d=lambda c:(c+c,c-1) #lambda input (a) colon :(logic with input)(a+1,a-2,a+3)
r=list(map(d,[1,2,3,4]))# list (map (input(logic to be used (should be described bf4), list)))
print(r)


[(2, 0), (4, 1), (6, 2), (8, 3)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [ ]:
d=lambda x:(x+x,x-1)
list(map(d,[1,2,3,4,5,5,6]))

[(2, 0), (4, 1), (6, 2), (8, 3), (10, 4), (10, 4), (12, 5)]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [ ]:
j=lambda x:x % 2==0
list(filter(j,[1,2,3,4]))

[2, 4]

In [ ]:
name=['angie','roma','diana','alex','aron','liya']
def a_names(name):
  if name[0].lower()=='a':
    return True
  else:
    return False

cap=list(filter(a_names,name))
print(cap)


False

#### Using Lambda's with Filter()

In [ ]:
a=lambda x:True if x+5==10 else False
b=list(filter(a,[5,4,5,4]))
b

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [ ]:
import statistics as s
import numpy as np
a=[1,2,3,4,5]
b=np.mean(a)
print(b)
g=lambda x:x<b
r=lambda x:x<b
list(filter(g,[1,2,3,4,5]))

3.0


[1, 2]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [ ]:
from functools import reduce
red=lambda x,y:x+y
reduce(red,[1,2,3,4,5])

15

#### Using Lambda's with Reduce()

In [ ]:
grd=lambda x,y:x*y
reduce(grd,[5,4,3,2,1])

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [ ]:
grd=lambda x,y:x*y
reduce(grd,[5,4,3,2,1])


120

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [ ]:
a=lambda x,y:x*y
reduce(a,[5,4,3,2,1])


def fact_func(num):
  fact=1
  for i in range(1,num+1):
    fact=fact*i
  return fact




fact_func(5)

120

In [ ]:
ages = [5, 12, 17, 18, 24, 32]
a=lambda x:True if x>=18 else False
b=list(filter(a,ages))
b

[18, 24, 32]

In [ ]:
def addnum(nums):
  if nums<=1:
    print("addnum(1)=1")#is there any connection bw red add num n yellow
    return nums
  else:
    print(f"addnum{nums}={nums}+addnum({nums-1})")
    return nums+addnum(nums-1)

addnum(20)

addnum20=20+addnum(19)
addnum19=19+addnum(18)
addnum18=18+addnum(17)
addnum17=17+addnum(16)
addnum16=16+addnum(15)
addnum15=15+addnum(14)
addnum14=14+addnum(13)
addnum13=13+addnum(12)
addnum12=12+addnum(11)
addnum11=11+addnum(10)
addnum10=10+addnum(9)
addnum9=9+addnum(8)
addnum8=8+addnum(7)
addnum7=7+addnum(6)
addnum6=6+addnum(5)
addnum5=5+addnum(4)
addnum4=4+addnum(3)
addnum3=3+addnum(2)
addnum2=2+addnum(1)
addnum(1)=1


210

#### Writing a Factorial Function

In [ ]:
def fact(n):
  if n <= 1:
    return 1
  else:
    return n*fact(n-1) #return n*(n-1)

fact(5)


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [ ]:
def sub_num(n): #check
  if n<=1:
    return 1
  else:
    print(f"(sub_num{n})=({n})-sub_num({n-1})")
    return n-sub_num(n-1)

sub_num(20)


(sub_num20)=(20)-sub_num(19)
(sub_num19)=(19)-sub_num(18)
(sub_num18)=(18)-sub_num(17)
(sub_num17)=(17)-sub_num(16)
(sub_num16)=(16)-sub_num(15)
(sub_num15)=(15)-sub_num(14)
(sub_num14)=(14)-sub_num(13)
(sub_num13)=(13)-sub_num(12)
(sub_num12)=(12)-sub_num(11)
(sub_num11)=(11)-sub_num(10)
(sub_num10)=(10)-sub_num(9)
(sub_num9)=(9)-sub_num(8)
(sub_num8)=(8)-sub_num(7)
(sub_num7)=(7)-sub_num(6)
(sub_num6)=(6)-sub_num(5)
(sub_num5)=(5)-sub_num(4)
(sub_num4)=(4)-sub_num(3)
(sub_num3)=(3)-sub_num(2)
(sub_num2)=(2)-sub_num(1)


10

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [ ]:
def even_num(n):
  for i in range(n):
    if i%2==0:
      yield i

result=even_num(20)
print(list(result))


[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
def gen(num):
  while num > 0:
    yield num**2
    num-=1

result=gen(5)
print(list(result))


[25, 16, 9, 4, 1]


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

result=filter(lambda a:len(a)>2,places)
list(result)

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

# res=sorted(author,key=lambda x:x.split()[-1].lower())
author.sort(key=lambda x:x.split()[-1].lower())
print(author)
# print(res)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


In [ ]:
d=lambda c:(c+c,c-1) #lambda input (a) colon :(logic with input)(a+1,a-2,a+3)
r=list(map(d,[1,2,3,4]))# list (map (input(logic to be used (should be described bf4), list)))
print(r)

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32 #y not key here
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
d=(lambda a:(a[0],a[1]*(9/5)+32))
result=list(map(d,places))
print(list(result))


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) =>
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [ ]:
def fibonacci(num):
  num1,num2=0,1
  print(num1,num2)
  for i in range(2,num+1):
    n3=num1+num2
    num1=num2
    num2=n3
    print(n3)

fibonacci(5)


0 1
1
2
3
5


In [ ]:
num=5
n1,n2=0,1
print(n1,n2)
for i in range (2,num+2):
  n3=n1+n2
  n1=n2
  n2=n3
  print(n3)

0 1
1
2
3
5
8
